### PR TITLE
[FIXED JENKINS-31068] Monitor does not detect when Tomcat URL encoding parameter rejects forward slashes in URL 

### PR DESCRIPTION
**Pull request by josiahhaswell**
See [JENKINS-31068](https://issues.jenkins-ci.org/browse/JENKINS-31068).

### Proposed changelog entries

* JENKINS-31068:  Proxy Monitor does not detect when Tomcat's URL escaping does not permit forward slashes.

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention
@reviewbybees

### CODE REVIEW

File: core/src/main/java/hudson/diagnosis/ReverseProxySetupMonitor.java [hunk 1]
```diff
@@ -27,16 +27,12 @@
 import hudson.Util;
 import hudson.model.AdministrativeMonitor;
 import org.jenkinsci.Symbol;
-import org.kohsuke.stapler.HttpRedirect;
-import org.kohsuke.stapler.HttpResponse;
-import org.kohsuke.stapler.HttpResponses;
-import org.kohsuke.stapler.QueryParameter;
+import org.kohsuke.stapler.*;
```
Comment (daniel-beck) [inline core/src/main/java/hudson/diagnosis/ReverseProxySetupMonitor.java]: Looks like your IDE needs some more configuration 😉

File: core/src/main/java/hudson/diagnosis/ReverseProxySetupMonitor.java [hunk 2]
```diff
@@ -100,5 +106,14 @@ public HttpResponse doAct(@QueryParameter String no) throws IOException {
     public String getDisplayName() {
         return Messages.ReverseProxySetupMonitor_DisplayName();
     }
+
+
+    private void checkEncoded(String encoded) {
+        if(!"a/b".equals(encoded)) {
+            throw HttpResponses.errorWithoutStack(404, "expected 'a/b'.  Got '" + encoded +
+                    "' (Are you in Apache Tomcat and have you run Tomcat with " +
```
Comment (daniel-beck) [inline core/src/main/java/hudson/diagnosis/ReverseProxySetupMonitor.java]: i18n?

File: core/src/main/java/hudson/diagnosis/ReverseProxySetupMonitor.java [hunk 3]
```diff
@@ -60,19 +56,27 @@ public boolean isActivated() {
...
         // May need to send an absolute URL, since handling of HttpRedirect with a relative URL does not currently honor X-Forwarded-Proto/Port at all.
-        String redirect = j.getRootUrl() + "administrativeMonitor/" + id + "/testForReverseProxySetup/" + (referer != null ? Util.rawEncode(referer) : "NO-REFERER") + "/";
+        String redirect = j.getRootUrl() + "administrativeMonitor/" + id + "/testForReverseProxySetup?rest=" + (referer != null ? Util.rawEncode(referer) : "NO-REFERER") + "&encoded=a%2Fb";
         LOGGER.log(Level.FINE, "coming from {0} and redirecting to {1}", new Object[] {referer, redirect});
         return new HttpRedirect(redirect);
     }
 
-    public void getTestForReverseProxySetup(String rest) {
+    public void doTestForReverseProxySetup(
+            StaplerRequest request,
```
Comment (daniel-beck) [inline core/src/main/java/hudson/diagnosis/ReverseProxySetupMonitor.java]: Unused args aren't necessary.

**Review by daniel-beck — COMMENTED** [review]
Seems reasonable if tested manually, including ensuring that the URL after toggling auto refresh does not result in false positive results.

**Comment by daniel-beck** [comment]
@josiahhaswell FYI https://wiki.jenkins-ci.org/display/JENKINS/GitHub+commit+messages, ideally use commit messages formatted like this:

> `[FIX JENKINS-12345] Short description of this commit`

File: core/src/main/java/hudson/diagnosis/ReverseProxySetupMonitor.java [hunk 4]
```diff
@@ -62,26 +67,35 @@ public boolean isActivated() {
     }
 
     public HttpResponse doTest() {
-        String referer = Stapler.getCurrentRequest().getReferer();
         Jenkins j = Jenkins.getInstance();
+        String referer = Stapler.getCurrentRequest().getReferer();
         // May need to send an absolute URL, since handling of HttpRedirect with a relative URL does not currently honor X-Forwarded-Proto/Port at all.
-        String redirect = j.getRootUrl() + "administrativeMonitor/" + id + "/testForReverseProxySetup/" + (referer != null ? Util.rawEncode(referer) : "NO-REFERER") + "/";
+        
+        String redirect = buildRedirect(referer, j);
```
Comment (jglick) [inline core/src/main/java/hudson/diagnosis/ReverseProxySetupMonitor.java]: I am finding it hard to follow the code here. What was an example value for `redirect` before and after this change?

File: core/src/test/java/hudson/diagnosis/ReverseProxySetupMonitorTest.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    
+    @Before
+    public void setUp() {
+        proxySetupMonitor = new ReverseProxySetupMonitor();
+    }
+    
+
+    @Test
+    public void ensureSplittingIsCorrect() throws UnsupportedEncodingException {
+        given(jenkins.getRootUrl()).willReturn("http://localhost:8080");
```
Comment (jglick) [inline core/src/test/java/hudson/diagnosis/ReverseProxySetupMonitorTest.java]: This is invalid; the root URL must end with a slash.

File: core/src/test/java/hudson/diagnosis/ReverseProxySetupMonitorTest.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        String referrer = proxySetupMonitor.buildRedirect("referrer", jenkins);
+        System.out.println(referrer);
+        String s = ReverseProxySetupMonitor.checkInput(URLDecoder.decode(referrer, "UTF-8"));
+        String expected = "http://localhost:8080administrativeMonitor/hudson.diagnosis.ReverseProxySetupMonitor/testForReverseProxySetup/referrer/";
+        assertThat(expected, is(s));
+    }
+
+    @Test(expected = HttpResponses.HttpResponseException.class)
+    public void ensureIncorrectlyEncodedURLFails() {
+        given(jenkins.getRootUrl()).willReturn("http://localhost:8080");
```
Comment (jglick) [inline core/src/test/java/hudson/diagnosis/ReverseProxySetupMonitorTest.java]: ditto

File: core/src/test/java/hudson/diagnosis/ReverseProxySetupMonitorTest.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.*;
+import static org.mockito.BDDMockito.given;
+
+/**
+ * Created by haswell on 8/11/17.
```
Comment (jglick) [inline core/src/test/java/hudson/diagnosis/ReverseProxySetupMonitorTest.java]: delete, and add license header

**Review by jglick — COMMENTED** [review]
I do not really understand what the code change is doing or why (would be helpful to explain exactly how the original monitor regressed and how this is correcting the problem), so I cannot meaningfully review the diff.

There is a test, but it is a mock-based unit test which does not prove effectiveness of the change; the original `ReverseProxySetupMonitorTest` (which used HtmlUnit to actually exercise the HTTP round-trips involved) apparently did not catch the regression, I think because it was only able to verify that there is _not_ a monitor displayed in the normal case using Jetty. If it is impractical to have an automated test based on our current frameworks reproduce the problem that Tomcat exhibits, which it probably is, then the PR should explain how this was tested manually, with enough detail that someone could recreate the test case. For example, [this Docker image](https://hub.docker.com/r/jglick/jenkins-demo-reverse-proxy/) lets you experiment with Jenkins behind an Apache reverse proxy; you could create something similar for Tomcat (if it is not already prepared somewhere) and describe how starting Jenkins in such an environment will display the expected monitor iff you have omitted the required Tomcat flag and you have applied this patch.

File: core/src/main/resources/hudson/diagnosis/Messages.properties [hunk 8]
```diff
@@ -6,4 +6,5 @@ HudsonHomeDiskUsageMonitor.DisplayName=Disk Usage Monitor
 
 NullIdDescriptorMonitor.DisplayName=Missing Descriptor ID
 ReverseProxySetupMonitor.DisplayName=Reverse Proxy Setup
+ReverseProxySetupMonitor.EncodedForwardSlashesDisallowed=Detected that encoded slashes are not allowed.  If you are running Tomcat, please be sure to run it with -Dorg.apache.tomcat.util.buf.UDecoder.ALLOW_ENCODED_SLASH=true
```
Comment (jglick) [inline core/src/main/resources/hudson/diagnosis/Messages.properties]: According to [this](https://stackoverflow.com/a/9933890/12916), Apache HTTPD also suffers from the same issue. Better to link to the [Jenkins wiki](https://wiki.jenkins.io/display/JENKINS/Jenkins+says+my+reverse+proxy+setup+is+broken) where the problem is described and solutions offered for various containers—as `doAct` already does.

**Comment by josiahhaswell** [comment]
So, the initial check would redirect to `administrativeMonitor/<className>/testForReverseProxySetup/a%2Fb/`, which if url slash encoding is disabled, would decode `a%2Fb` to `a/b` during URL processing, and so Stapler would try to resolve a sub-resource `a` on `testForReverseProxySetup`, which (always, now and then) fails.  When URL encoding is enabled, however, resource resolution succeeds up to `testForReverseProxySetup` and the correctly decoded path parameter is passed into the method, at which the other checks are applied.  The regression is that if no referrer is added during the redirect, no encoded URL portion will be sent along with it.  This ensures that an encoded fragment will always be passed along with the URL redirect.  The tests just verify that that is the case and that it's processed correctly.

File: core/src/main/java/hudson/diagnosis/ReverseProxySetupMonitor.java (hunk 4 above)
Comment (josiahhaswell) [inline core/src/main/java/hudson/diagnosis/ReverseProxySetupMonitor.java]: In the case of a referrer: 

`10.10.10.10/some/path/`

In the case of no referrer:
`NO-REFERRER`

Now it's:

Referrer: 

`10.10.10.10/some/path/~~no%2Fscheme~~`
No Referrer:
`NO-REFERRER/~~no%2Fscheme~~` // note that an encoded portion is always passed--this is the case that changed and introduced the regression

File: core/src/test/java/hudson/diagnosis/ReverseProxySetupMonitorTest.java (hunk 5 above)
Comment (josiahhaswell) [inline core/src/test/java/hudson/diagnosis/ReverseProxySetupMonitorTest.java]: I'll change that, but this is just to check the referrer construction.  None of the system tests will test this since this is not manifested in Jetty.

File: core/src/main/java/hudson/diagnosis/ReverseProxySetupMonitor.java (hunk 4 above)
Comment (josiahhaswell) [inline core/src/main/java/hudson/diagnosis/ReverseProxySetupMonitor.java]: (the `~~` are used as a delimiter because they're non-reserved special characters that won't appear in a referrer URL and cause extraction to fail)

**Comment by josiahhaswell** [comment]
Steps to manually test:

1.) Install Tomcat (any version works)
2.) Run Tomcat without ` -Dorg.apache.tomcat.util.buf.UDecoder.ALLOW_ENCODED_SLASH=true`
3.) Navigate to `http://localhost:8080/administrativeMonitor/hudson.diagnosis.ReverseProxySetupMonitor/test`
4.) Receive 400 error (same behavior as prior to regression)
5.) Restart Tomcat with` -Dorg.apache.tomcat.util.buf.UDecoder.ALLOW_ENCODED_SLASH=true`
6.) Navigate to `http://localhost:8080/administrativeMonitor/hudson.diagnosis.ReverseProxySetupMonitor/test`
7.) Receive 200 OK

File: core/src/main/resources/hudson/diagnosis/Messages.properties [hunk 9]
```diff
@@ -6,5 +6,6 @@ HudsonHomeDiskUsageMonitor.DisplayName=Disk Usage Monitor
 
 NullIdDescriptorMonitor.DisplayName=Missing Descriptor ID
 ReverseProxySetupMonitor.DisplayName=Reverse Proxy Setup
-ReverseProxySetupMonitor.EncodedForwardSlashesDisallowed=Detected that encoded slashes are not allowed.  If you are running Tomcat, please be sure to run it with -Dorg.apache.tomcat.util.buf.UDecoder.ALLOW_ENCODED_SLASH=true
+ReverseProxySetupMonitor.EncodedForwardSlashesDisallowed=Detected that encoded slashes are not allowed.  If you are running Tomcat, please be sure to run it with -Dorg.apache.tomcat.util.buf.UDecoder.ALLOW_ENCODED_SLASH=true\n\
+  more info at: https://wiki.jenkins.io/display/JENKINS/Jenkins+says+my+reverse+proxy+setup+is+broken
```
Comment (jglick) [inline core/src/main/resources/hudson/diagnosis/Messages.properties]: I would recommend linking only to the wiki, as `doAct` did, and keep all container-specific details there.

**Comment by jglick** [comment]
> Navigate to `http://localhost:8080/administrativeMonitor/hudson.diagnosis.ReverseProxySetupMonitor/test`

Clearer to go to `/manage` and look at whether a warning is displayed or not.

**Review by jglick — APPROVED** [review]
Do not follow the details but sounds like it has been manually tested, so good enough.

File: core/src/main/resources/hudson/diagnosis/Messages.properties (hunk 9 above)
Comment (daniel-beck) [inline core/src/main/resources/hudson/diagnosis/Messages.properties]: Seems I forgot to rewrite a URL to `jenkins.io/redirect`. Would be great if this linked to `https://jenkins.io/redirect/troubleshooting/reverse-proxy-broken` instead; I'll set up the redirect on jenkins.io if implemented.

**Comment by josiahhaswell** [comment]
I found another issue with this while I was testing in Tomcat behind Apache HTTPD.  I need to fix that, and I think it'll require another administrative monitor--ReverseProxySetupMonitor mostly checks something else besides the encoding now, and I can't really reconcile it to this check.

**Review by jglick — APPROVED** [review]

**Review by oleg-nenashev — APPROVED** [review]
The explanation sounds reasonable to me, though I am not 100% confident about that. 🐝 and @reviewbybees done, but needs some feedback from @jenkinsci/code-reviewers imho

**Review by daniel-beck — APPROVED** [review]

**Comment by daniel-beck** [comment]
Needs squashing.
